docs/docs/reference/other-new-features/explicit-nulls.md

```diff
@@ -0,0 +1,296 @@
+    ```
+
+    Suppose we have a `BoxFactory[String]`. Notice that calling `makeBox()` on it returns a `Box[String]|JavaNull`, not
+    a `Box[String|JavaNull]|JavaNull`, because of rule 4. This seems at first glance unsound ("What if the box itself
+    has `null` inside?"), but is sound because calling `get()` on a `Box[String]` returns a `String|JavaNull`, as per
+    rule 3.
+
+    Notice that for rule 4 to be correct we need to patch _all_ Java-defined classes that transitively appear in the
+    argument or return type of a field or method accessible from the Scala code being compiled. Absent crazy reflection
+    magic, we think that all such Java classes _must_ be visible to the Typer in the first place, so they will be patched.
+
+  * Rule 5 is needed because Java code might use a generic that's defined in Scala as opposed to Java.
+    ```scala
+    class BoxFactory<T> { Box<T> makeBox(); } // Box is Scala defined
+    ==>
+    class BoxFactory[T] { def makeBox(): Box[T|JavaNull]|JavaNull }
+    ```
+
+    In this case, since `Box` is Scala-defined, `nf` is applied to the type argument `T`, so rule 3 applies and we get
+    `Box[T|JavaNull]|JavaNull`. This is needed because our nullability function is only applied (modularly) to the Java
+    classes, but not to the Scala ones, so we need a way to tell `Box` that it contains a nullable value.
+
+  * Rules 6, 7, and 8 just recurse structurally on the components of the type.
+    The handling of unions and intersections in the compiler is a bit more involved than the presentation above.
+    Specifically, the implementation makes sure to add `| Null` only at the top level of a type:
+    e.g. `nf(A & B) = (A & B) | JavaNull`, as opposed to `(A | JavaNull) & (B | JavaNull)`.
+
+### JavaNull
+To enable method chaining on Java-returned values, we have a special `JavaNull` alias
+```scala
+type JavaNull = Null
+```
+
+`JavaNull` behaves just like `Null`, except it allows (unsound) member selections:
+```scala
+// Assume someJavaMethod()'s original Java signature is
+// String someJavaMethod() {}
+val s2: String = someJavaMethod().trim().substring(2).toLowerCase() // unsound
+```
+
+Here, all of `trim`, `substring` and `toLowerCase` return a `String|JavaNull`.
+The Typer notices the `JavaNull` and allows the member selection to go through.
+However, if `someJavaMethod` were to return `null`, then the first member selection
+would throw a `NPE`.
+
+Without `JavaNull`, the chaining becomes too cumbersome
+```scala
+val ret = someJavaMethod()
+val s2 = if (ret != null) {
+  val tmp = ret.trim()
+  if (tmp != null) {
+    val tmp2 = tmp.substring(2)
+    if (tmp2 != null) {
+      tmp2.toLowerCase()
+    }
+  }
+}
+// Additionally, we need to handle the `else` branches.
+```
+
+## Binary Compatibility
+
+Our strategy for binary compatibility with Scala binaries that predate explicit nulls
+is to leave the types unchanged and be compatible but unsound.
```

Comment:
If I understand correclty, this section applies not only to Scala libraries compiled before explicit nulls were available, but also to new libraries compiled without `-Yexplicit-nulls` - which will probably be the majority for a while. It would be good to mention this explicitly.